xrootd/encoder/encoder.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package encoder // import "go-hep.org/x/hep/xrootd/encoder"
+
+import (
+	"encoding/binary"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+// MarshalRequest encodes request body alongside with request and stream ids
+func MarshalRequest(requestID uint16, streamID protocol.StreamID, requestBody interface{}) ([]byte, error) {
+	requestHeader := make([]byte, 4)
+
+	copy(requestHeader, streamID[:])
+	binary.BigEndian.PutUint16(requestHeader[2:], requestID)
+
+	b, err := Marshal(requestBody)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(requestHeader, b...), nil
+}
+
+// Marshal encodes structure to the bytes following the XRootd protocol specification.
+// Fields are encoded in the same order as they are specified in the struct definition.
+// Each field is encoded in network byte order (BigEndian) without any alignment or padding.
+// Slices and arrays of uint8 are binary copied without further encoding.
+// Supported types are: uint8, uint16, int32, int64, slices and arrays of uint8.
+func Marshal(x interface{}) ([]byte, error) {
+	v := reflect.ValueOf(x)
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+
+	if v.Kind() != reflect.Struct {
+		return nil, errors.Errorf("Cannot marshal %s, struct is expected", v.Kind())
+	}
+
+	dataSize, err := calculateSizeForMarshaling(v)
+	if err != nil {
+		return nil, err
+	}
+
+	data := make([]byte, dataSize)
+	pos := 0
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		fieldSize := 0
+		switch field.Kind() {
+		case reflect.Uint8:
+			fieldSize = 1
+			data[pos] = uint8(field.Uint())
+		case reflect.Uint16:
+			fieldSize = 2
+			binary.BigEndian.PutUint16(data[pos:pos+fieldSize], uint16(field.Uint()))
+		case reflect.Int32:
+			fieldSize = 4
+			binary.BigEndian.PutUint32(data[pos:pos+fieldSize], uint32(field.Int()))
+		case reflect.Int64:
+			fieldSize = 8
+			binary.BigEndian.PutUint64(data[pos:pos+fieldSize], uint64(field.Int()))
+		case reflect.Slice:
+			elementKind := field.Type().Elem().Kind()
+			if elementKind != reflect.Uint8 {
+				return nil, errors.Errorf("Cannot marshal slice of %s, only slices of uint8 are supported", elementKind)
+			}
+
+			fieldSize = field.Len()
+			reflect.Copy(reflect.ValueOf(data[pos:pos+fieldSize]), field)
+		case reflect.Array:
+			elementKind := field.Type().Elem().Kind()
+			if elementKind != reflect.Uint8 {
+				return nil, errors.Errorf("Cannot marshal array of %s, only arrays of uint8 are supported", elementKind)
+			}
+
+			fieldSize = field.Len()
+			reflect.Copy(reflect.ValueOf(data[pos:pos+fieldSize]), field)
```

Comment:
ditto.
except that would probably look like:
```go
v := field.Slice(0, field.Len()).Interface().([]byte)
fieldSize = copy(data[pos:], v)
```
(untested)

Reply:
I have got `panic: reflect.Value.Slice: slice of unaddressable array` with this snippet. Reduced duplication anyway, thanks.

Reply:
ah... right.
something like this could work:
```go
case reflect.Array:
    // Create slice from array
    var n = reflect.New(v.Type())
    n.Elem().Set(v)
    v = reflect.Indirect(n).Slice(0, v.Len())
    fallthrough
case reflect.Slice:
```
(but it copies the whole array...)